pkg/lib/authn/identity/ldap/provider.go

```diff
@@ -100,3 +109,35 @@ func sortIdentities(is []*identity.LDAP) {
 		return is[i].CreatedAt.Before(is[j].CreatedAt)
 	})
 }
+
+func (p *Provider) CreateNormalizedIdentitySpecFromLDAPEntry(serverConfig *config.LDAPServerConfig, entry *ldap.Entry) (*identity.Spec, error) {
+	userIDAttributeName := serverConfig.UserIDAttributeName
+	userIDAttributeValue := entry.GetRawAttributeValue(userIDAttributeName)
+
+	claims := map[string]interface{}{}
+	if v := entry.GetAttributeValue(ldap.AttributeMail.Name); v != "" {
+		claims[string(model.ClaimEmail)] = v
+	}
+	if v := entry.GetAttributeValue(ldap.AttributeMobile.Name); v != "" {
+		claims[string(model.ClaimPhoneNumber)] = v
+	}
+	if v := entry.GetAttributeValue(ldap.AttributeUID.Name); v != "" {
+		claims[string(model.ClaimPreferredUsername)] = v
+	}
+
+	err := p.StandardAttributesNormalizer.Normalize(claims)
+	if err != nil {
+		claims = map[string]interface{}{}
+	}
+
+	return &identity.Spec{
+		Type: model.IdentityTypeLDAP,
+		LDAP: &identity.LDAPSpec{
+			ServerName:           serverConfig.Name,
+			UserIDAttributeName:  userIDAttributeName,
+			UserIDAttributeValue: userIDAttributeValue,
+			Claims:               claims,
+			RawEntryJSON:         entry.ToJSON(),
+		},
+	}, nil
+}
```

Comment:
Don't pollute ldap identity provider

Reply:
From what I know, identity provider is dealing with identity

Reply:
Actually I think this is a good place to put this code. We need a place to convert `*ldap.Entry` to a `*identity.Spec`. Since the conversion involves Authgear-specific logic (when we have attribute mapping), LDAP client is not a good candidate to host this code.